hl_extractor/job_calc.py

```diff
@@ -0,0 +1,241 @@
+    job = db.dataset_eval.get_job(jobid)
+    dataset = db.dataset.get(job["dataset_id"])
+    model_name = db.dataset._slugify(dataset["name"])
+    model_version = "v1" # TODO: This has to be a parameter (from profile?)
+
+    model_id = db.data._get_model_id(model_name, model_version)
+    if model_id is None:
+        model_id = db.data.add_model(model_name, model_version)
+
+    return model_id
+
+
+def main(num_threads, profile, dataset_job_id):
+    print("High-level extractor daemon starting with %d threads" % num_threads)
+    sys.stdout.flush()
+    build_sha1 = get_build_sha1(HIGH_LEVEL_EXTRACTOR_BINARY)
+    create_profile(profile, PROFILE_CONF, build_sha1)
+    db.init_db_engine(config.SQLALCHEMY_DATABASE_URI)
+
+    model_id = get_model_from_eval(dataset_job_id)
+    includes = load_includes_from_eval(dataset_job_id)
+
+    num_processed = 0
+
+    pool = {}
+    docs = []
+    while True:
+        # Check to see if we need more database rows
+        if len(docs) == 0:
+            # Fetch more rows from the DB
+            docs = db.data.get_unprocessed_highlevel_documents_for_model(model_id, includes)
+
+            # We will fetch some rows that are already in progress. Remove those.
+            in_progress = pool.keys()
+            filtered = []
+            for mbid, doc, id in docs:
+                if id not in in_progress:
+                    filtered.append((mbid, doc, id))
+            docs = filtered
+
+        if len(docs):
+            # Start one document
+            mbid, doc, id = docs.pop()
+            th = HighLevel(mbid, doc, id)
+            th.start()
+            print("start %s" % mbid)
+            sys.stdout.flush()
+            pool[id] = th
+
+        # If we're at max threads, wait for one to complete
+        while True:
+            if len(pool) == 0 and len(docs) == 0:
+                if num_processed > 0:
+                    print("processed %s documents, none remain. Sleeping." % num_processed)
+                    sys.stdout.flush()
+                num_processed = 0
+                # Let's be nice and not keep any connections to the DB open while we nap
+                # TODO: Close connections when we're sleeping
+                sleep(SLEEP_DURATION)
+
+            for id in pool.keys():
+                if not pool[id].is_alive():
+
+                    # Fetch the data and clean up the thread object
+                    hl_data = pool[id].get_data()
+                    ll_id = pool[id].get_ll_id()
+                    pool[id].join()
+                    del pool[id]
+
+                    try:
+                        jdata = json.loads(hl_data)
+                    except ValueError:
+                        print("error %s: Cannot parse result document" % mbid)
+                        print(hl_data)
+                        sys.stdout.flush()
+                        jdata = {}
+
+                    db.data.write_high_level(mbid, ll_id, jdata, build_sha1)
```

Comment:
If `jdata` is empty, we need to work out how to add a dummy highlevel item for the combination of `ll_id` and `module`, so that when we ask for the next lot of data this failed example is not here.